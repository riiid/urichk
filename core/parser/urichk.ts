import {
  createRecursiveDescentParser,
  Pattern,
  RecursiveDescentParser,
  Span,
  Token,
} from "https://deno.land/x/pbkit@v0.0.39/core/parser/recursive-descent-parser.ts";
import * as ast from "../ast/index.ts";

export interface ParseResult<T = ast.Urichk> {
  ast: T;
  parser: RecursiveDescentParser;
}

export function parse(text: string): ParseResult {
  const parser = createRecursiveDescentParser(text);
  const rules = acceptStatements(parser, [acceptRule]);
  return { ast: rules, parser };
}

interface AcceptStatementFn<T extends ast.StatementBase> {
  (
    parser: RecursiveDescentParser,
    leadingComments: ast.CommentGroup[],
  ): T | undefined;
}
function acceptStatement<T extends ast.StatementBase>(
  parser: RecursiveDescentParser,
  acceptStatementFn: AcceptStatementFn<T>,
) {
  const { commentGroups: leadingComments } = skipWsAndSweepComments(parser);
  const statement = acceptStatementFn(
    parser,
    leadingComments,
  );
  return statement;
}

function acceptStatements<T extends ast.StatementBase>(
  parser: RecursiveDescentParser,
  acceptStatementFns: AcceptStatementFn<T>[],
) {
  const statements: T[] = [];
  statements:
  while (true) {
    const { commentGroups: leadingComments } = skipWsAndSweepComments(parser);
    for (const acceptStatementFn of acceptStatementFns) {
      const statement = acceptStatementFn(
        parser,
        leadingComments,
      );
      if (statement) {
        statements.push(statement);
        continue statements;
      }
    }
    break;
  }
  return statements;
}

function acceptRule(
  parser: RecursiveDescentParser,
  leadingComments: ast.CommentGroup[],
): ast.Rule | undefined {
  const head = acceptStatement(parser, acceptHead);
  if (!head) return;
  skipWsAndComments(parser);
  const tail = acceptStatement(parser, acceptTail);
  if (!tail) return;
  const trailingComments = acceptTrailingComments(parser);
  return {
    ...mergeSpans([
      head,
      tail,
      leadingComments,

      trailingComments,
    ]),
    head,
    tail,
    leadingComments,

    trailingComments,
  };
}

// @TODO: https://datatracker.ietf.org/doc/html/rfc3986#section-3
function acceptHead(
  parser: RecursiveDescentParser,
  leadingComments: ast.CommentGroup[],
): ast.Head | undefined {
  const schemeOrAuthority = acceptSchemeOrAuthority(parser);
  if (schemeOrAuthority) {
    const { scheme, authority } = schemeOrAuthority;
    const path = acceptPath(parser);
    const trailingComments = acceptTrailingComments(parser);
    return {
      ...mergeSpans([
        scheme,
        authority,
        path,
        leadingComments,
      ]),
      scheme,
      authority,
      path,
      leadingComments,
      trailingComments,
    };
  } else {
    const path = acceptPath(parser);
    const trailingComments = acceptTrailingComments(parser);
    if (!path) return;
    return {
      ...mergeSpans([
        path,
        leadingComments,
      ]),
      path,
      leadingComments,
      trailingComments,
    };
  }
  function acceptSchemeOrAuthority(parser: RecursiveDescentParser) {
    const loc = parser.loc;
    return acceptPatternAndThen(schemePattern, (token) => {
      if (!token) return;
      const isSchemeExist = parser.accept(":");
      if (!isSchemeExist) parser.loc = loc;
      return isSchemeExist
        ? { scheme: token, authority: acceptAuthority(parser) }
        : { authority: acceptAuthority(parser) };
    })(parser);
  }
}

function acceptAuthority(
  parser: RecursiveDescentParser,
): ast.Authority | undefined {
  const slash = parser.accept(/^\/\//);
  const userinfoAndHost = acceptUserinfoAndHost(parser);
  if (!userinfoAndHost) return;
  const { userinfo, host } = userinfoAndHost;
  const port = parser.accept(":") && parser.accept(identPattern);
  return {
    ...mergeSpans([slash, userinfo, host, port]),
    userinfo,
    host,
    port,
  };
  function acceptUserinfoAndHost(parser: RecursiveDescentParser) {
    return acceptPatternAndThen(identPattern, (token) => {
      if (!token) return;
      const isUserinfoExist = parser.accept("@");
      return isUserinfoExist
        ? { userinfo: token, host: parser.expect(identPattern) }
        : { host: token };
    })(parser);
  }
}

function acceptPath(parser: RecursiveDescentParser): ast.Path | undefined {
  const fragments: ast.PathFragment[] = [];
  while (parser.accept("/")) {
    const fragment = acceptPathFragment(parser);
    if (fragment) fragments.push(fragment);
  }
  if (fragments.length === 0) return;
  return {
    ...mergeSpans(fragments),
    fragments,
  };
}

// @TODO: https://datatracker.ietf.org/doc/html/rfc3986#section-3.3
function acceptPathFragment(
  parser: RecursiveDescentParser,
): ast.PathFragment | undefined {
  const isParamPathFragment = Boolean(parser.accept("["));
  const path = parser.accept(identPattern);
  isParamPathFragment && parser.expect("]");
  if (!path) return;
  return { type: isParamPathFragment ? "param" : "static", ...path };
}

function acceptTail(
  parser: RecursiveDescentParser,
  leadingComments: ast.CommentGroup[],
): ast.Tail | undefined {
  const openParens = parser.expect("{");
  const rules = acceptTailRules(parser);
  const closeParens = parser.expect("}");
  const trailingComments = acceptTrailingComments(parser);
  return {
    ...mergeSpans([
      openParens,
      closeParens,
      leadingComments,
      trailingComments,
    ]),
    openParens,
    rules,
    closeParens,
    leadingComments,
    trailingComments,
  };
}

function acceptTailRules(parser: RecursiveDescentParser): ast.TailRule[] {
  const tailRules: ast.TailRule[] = [];
  while (true) {
    const tailRule = acceptStatement(parser, acceptTailRule);
    if (!tailRule) break;
    tailRules.push(tailRule);
    skipWsAndComments(parser);
  }
  return tailRules;
}

function acceptTailRule(
  parser: RecursiveDescentParser,
  leadingComments: ast.CommentGroup[],
): ast.TailRule | undefined {
  const tailType = parser.accept(/^(\?|#)/) as ast.Token<"?" | "#">;
  if (!tailType) return;
  skipWsAndComments(parser);
  const matchType = parser.accept(/^(form|match)/) as ast.Token<
    "match" | "form"
  >;
  if (!matchType) return;
  skipWsAndComments(parser);
  switch (matchType.text) {
    case "form": {
      const openParens = parser.accept("{");
      // skipWsAndComments(parser);
      const pattern = acceptTailRuleFormPatternRules(parser);
      skipWsAndComments(parser);
      const closeParens = parser.accept("}");
      return {
        openParens,
        closeParens,
        tailType,
        matchType,
        pattern,
        leadingComments,
        trailingComments: acceptTrailingComments(parser),
      } as ast.TailRuleForm; // @FIXME: typing!
    }
    case "match": {
      return {
        tailType,
        matchType,
        pattern: acceptTailRuleMatchPatternRules(parser),
        leadingComments,
        trailingComments: acceptTrailingComments(parser),
      } as ast.TailRuleMatch;
    }
  }
}

function acceptTailRuleFormPatternRules(
  parser: RecursiveDescentParser,
): ast.TailRuleFormPatternRule[] {
  const rules: ast.TailRuleFormPatternRule[] = [];
  while (!parser.try(/^\s*}/)) {
    const { commentGroups: leadingComments } = skipWsAndSweepComments(parser);
    const key = choice<ast.Token>([acceptIdent, acceptStrLit])(
      parser,
    ) as ast.Key | undefined;
    if (!key) break;
    skipWsAndComments(parser);
    const array = parser.accept("[]");
    skipWsAndComments(parser);
    const equal = parser.accept("=");
    if (equal) {
      const values: ast.TailRulePatternValue[] =
        acceptTailRuleMatchPatternRules(parser);
      const trailingComments = acceptTrailingComments(parser);
      rules.push({
        ...mergeSpans([
          key,
          ...values,
          leadingComments,

          trailingComments,
        ]),
        key,
        array,
        value: values,
        leadingComments,

        trailingComments,
      });
    } else {
      const trailingComments = acceptTrailingComments(parser);
      rules.push({
        ...mergeSpans([
          key,
          leadingComments,

          trailingComments,
        ]),
        key,
        array,
        leadingComments,

        trailingComments,
      });
    }
  }
  return rules;
}

function acceptTailRuleMatchPatternRules(
  parser: RecursiveDescentParser,
): ast.TailRuleMatchPatternRule[] {
  let loc = parser.loc;
  const rules: ast.TailRuleMatchPatternRule[] = [];
  parser.accept("|");
  do {
    const { commentGroups: leadingComments } = skipWsAndSweepComments(parser);
    const rule = choice([acceptRegex, acceptIdent, acceptStrLit])(
      parser,
    ) as ast.TailRuleMatchPatternRule;
    rules.push({
      ...rule,
      leadingComments,
      trailingComments: acceptTrailingComments(parser),
    });
    loc = parser.loc;
    skipWsAndComments(parser);
  } while (parser.accept("|"));
  parser.loc = loc;
  return rules;
}

function acceptIdent(parser: RecursiveDescentParser) {
  const ident = parser.accept(identPattern);
  if (!ident) return;
  return { ...ident, type: "id" };
}

function acceptStrLit(parser: RecursiveDescentParser) {
  const strLit = parser.accept(strLitPattern);
  if (!strLit) return;
  return { ...strLit, type: "string" };
}

function acceptRegex(parser: RecursiveDescentParser) {
  const regex = parser.accept(regexPattern);
  if (!regex) return;
  return { ...regex, type: "regex" };
}

function mergeSpans(spans: (undefined | Span | (undefined | Span)[])[]): Span {
  let start = Infinity;
  let end = -Infinity;
  for (let i = 0; i < spans.length; ++i) {
    if (spans[i] == null) continue;
    const span = Array.isArray(spans[i])
      ? mergeSpans(spans[i] as Span[])
      : spans[i] as Span;
    start = Math.min(start, span.start);
    end = Math.max(end, span.end);
  }
  return { start, end };
}

interface AcceptFn<T> {
  (parser: RecursiveDescentParser): T | undefined;
}

function acceptPatternAndThen<T>(
  pattern: Pattern,
  then: (token: Token | undefined) => T,
): AcceptFn<T> {
  return function accept(parser) {
    const token = parser.accept(pattern);
    return then(token);
  };
}

function choice<T>(acceptFns: AcceptFn<T>[]): AcceptFn<T> {
  return function accept(parser) {
    for (const acceptFn of acceptFns) {
      const node = acceptFn(parser);
      if (node) return node;
    }
  };
}

function many<T>(parser: RecursiveDescentParser, acceptFn: AcceptFn<T>): T[] {
  const nodes: T[] = [];
  let node: ReturnType<typeof acceptFn>;
  while (node = acceptFn(parser)) nodes.push(node);
  return nodes;
}

const whitespacePattern = /^\s+/;
const whitespaceWithoutNewlinePattern =
  /^[ \f\t\v\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000\ufeff]+/;
const newlinePattern = /^\r?\n/;
const identPattern = /^(?:[a-zA-Z0-9\-._~]|%[0-9a-fA-F]{2})+/;
const strLitPattern = /^'.*?'/;
const multilineCommentPattern = /^\/\*(?:.|\r?\n)*?\*\//;
const documentCommentPattern = /^\/\*\*(?:.|\r?\n\s*\*)*(\r?\n\s*)\*\//;
const singlelineCommentPattern = /^\/\/.*(?:\r?\n|$)/;
const schemePattern = /^[a-z][a-z0-9\+-\.]*/i;
const regexPattern =
  /^\/(?:(?![*+?])(?:[^\r\n\[/\\]|\\.|\[(?:[^\r\n\]\\]|\\.)*\])+)\/[a-z]*/;

function acceptSpecialToken<TType extends string>(
  parser: RecursiveDescentParser,
  type: TType,
  pattern: Pattern = identPattern,
): (Token & { type: TType }) | undefined {
  const token = parser.accept(pattern);
  if (!token) return;
  return { type, ...token };
}

function acceptCommentGroup(
  parser: RecursiveDescentParser,
): ast.CommentGroup | undefined {
  const loc = parser.loc;
  const comments: ast.Comment[] = [];
  while (true) {
    const whitespace = parser.accept(whitespaceWithoutNewlinePattern);
    if (whitespace) continue;
    const documentComment = acceptSpecialToken(
      parser,
      "document-comment",
      documentCommentPattern,
    );
    if (documentComment) {
      comments.push(documentComment);
      continue;
    }
    const multilineComment = acceptSpecialToken(
      parser,
      "multiline-comment",
      multilineCommentPattern,
    );
    if (multilineComment) {
      comments.push(multilineComment);
      continue;
    }
    const singlelineComment = acceptSpecialToken(
      parser,
      "singleline-comment",
      singlelineCommentPattern,
    );
    if (singlelineComment) {
      comments.push(singlelineComment);
      continue;
    }
    break;
  }
  if (comments.length < 1) {
    parser.loc = loc;
    return;
  }
  return {
    ...mergeSpans(comments),
    type: "comment-group",
    comments,
  };
}

function acceptTrailingComments(
  parser: RecursiveDescentParser,
): ast.CommentGroup[] {
  const loc = parser.loc;
  const comments: ast.Comment[] = [];
  while (true) {
    const whitespace = parser.accept(whitespaceWithoutNewlinePattern);
    if (whitespace) continue;
    const newline = parser.accept(newlinePattern);
    if (newline) break;
    const documentComment = acceptSpecialToken(
      parser,
      "document-comment",
      documentCommentPattern,
    );
    if (documentComment) {
      comments.push(documentComment);
      continue;
    }
    const multilineComment = acceptSpecialToken(
      parser,
      "multiline-comment",
      multilineCommentPattern,
    );
    if (multilineComment) {
      comments.push(multilineComment);
      continue;
    }
    const singlelineComment = acceptSpecialToken(
      parser,
      "singleline-comment",
      singlelineCommentPattern,
    );
    if (singlelineComment) {
      comments.push(singlelineComment);
      break;
    }
    break;
  }
  if (comments.length < 1) {
    parser.loc = loc;
    return [];
  }
  return [{
    ...mergeSpans(comments),
    type: "comment-group",
    comments,
  }];
}

interface SkipWsAndSweepCommentsResult {
  commentGroups: ast.CommentGroup[];
  trailingNewline: boolean;
}
function skipWsAndSweepComments(
  parser: RecursiveDescentParser,
): SkipWsAndSweepCommentsResult {
  const commentGroups: ast.CommentGroup[] = [];
  let trailingNewline = false;
  parser.accept(whitespacePattern);
  while (true) {
    const commentGroup = acceptCommentGroup(parser);
    if (commentGroup) {
      commentGroups.push(commentGroup);
      trailingNewline = false;
      continue;
    }
    const whitespace = parser.accept(whitespaceWithoutNewlinePattern);
    if (whitespace) continue;
    const newline = parser.accept(newlinePattern);
    if (newline) {
      trailingNewline = true;
      continue;
    }
    break;
  }
  return {
    commentGroups,
    trailingNewline,
  };
}

function skipWsAndComments(parser: RecursiveDescentParser): undefined {
  while (true) {
    const whitespace = parser.accept(whitespacePattern);
    if (whitespace) continue;
    const multilineComment = acceptSpecialToken(
      parser,
      "multiline-comment",
      multilineCommentPattern,
    );
    if (multilineComment) continue;
    const singlelineComment = acceptSpecialToken(
      parser,
      "singleline-comment",
      singlelineCommentPattern,
    );
    if (singlelineComment) continue;
    break;
  }
  return;
}
