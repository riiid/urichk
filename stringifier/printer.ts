export interface Printer {
  indent(): void;
  dedent(): void;
  print(text: string): void;
  printIndent(): void;
  println(line: string): void;
  done(): string;
}

export interface CreatePrinterConfig {
  space?: string;
}
export function createPrinter(config?: CreatePrinterConfig): Printer {
  const { space = "  " } = { ...config };
  const buffer: string[] = [];
  let _indent = 0;
  return {
    indent() {
      ++_indent;
    },
    dedent() {
      --_indent;
    },
    print(text) {
      buffer.push(text);
    },
    printIndent() {
      buffer.push(space.repeat(_indent));
    },
    println(line) {
      buffer.push(space.repeat(_indent));
      buffer.push(line);
      buffer.push("\n");
    },
    done() {
      return buffer.join("");
    },
  };
}
