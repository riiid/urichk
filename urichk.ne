@preprocessor urichk

@{%
  import moo from "https://deno.land/x/moo@0.5.1.1/index.ts";
  const commonTokenRules = {
    num: /[0-9]+/,
    ln: { match: /\r?\n/, lineBreaks: true },
    ws: /[ \t]+/,
    id: /(?:[a-zA-Z0-9\-._~]|%[0-9a-fA-F]{2})+/,
    sc: /\/\/[^\r\n]*\r?\n/,
    mc: /\/\*(?:\*(?!\/)|[^*])*\*\//,
  };
  const lexer = moo.states({
    root: {
      ...commonTokenRules,
      lcb: { match: '{', push: 'block' },
      reserved: /[:/?#\[\]@]/,
      permitted: /[!$&'()*+,;=]/,
    },
    block: {
      ...commonTokenRules,
      lcb: { match: '{', push: 'block' },
      rcb: { match: '}', pop: 1 },
      string: /'.*?'/,
      symbol: /[?#:'=|\[\]]/,
      regex: /\/(?:(?![*+?])(?:[^\r\n\[/\\]|\\.|\[(?:[^\r\n\]\\]|\\.)*\])+)\/[a-z]*/,
    },
  });
%}
@lexer lexer

urichk -> rule:* _ {% id %}
_ -> (%ws | %ln | %sc | %mc):* {%
  ([tokens]) => {
    let result = undefined;
    for (const [token] of tokens) {
      if (token.type === 'mc') result = token;
    }
    return result;
  }
%}
ident => %id {% id %} | %num {% id %}

rule -> _ head _ tail {% ([comment, head, , tail]) => ({ comment, head, tail }) %}

head ->
    scheme authority path:? {% parseHead %}
  | scheme:? authority:? path {% parseHead %}
@{%
  const parseHead = ([scheme, authority, path]) => ({
    scheme,
    authority,
    path,
  });
%}

tail -> "{" (_ tail_rule):* _ "}" {% ([, rules]) => rules.map(([comment, rule]) => ({ comment, ...rule })) %}

scheme -> ident ":" {% id %}
authority ->
  ("/" "/"):? userinfo:? ident port:?
  {% ([, userinfo, host, port]) => ({
    userinfo,
    host,
    port,
  }) %}
userinfo -> ident "@" {% id %}
host -> ident {% id %}
port -> ":" %num {% ([, port]) => port %}
path ->
    "/" {% () => [] %}
  | ("/" path_fragment):+ "/":? {% ([fragments]) => fragments.map(([, fragment]) => fragment) %}
path_fragment ->
    ident {% ([name]) => ({ type: 'static', name }) %}
  | "[" _ ident _ "]" {% ([, , name]) => ({ type: 'param', name }) %}

tail_rule ->
    tail_rule_match {% id %}
  | tail_rule_form {% id %}
tail_rule_tail_type ->
    "?" {% id %}
  | "#" {% id %}
tail_rule_match ->
  tail_rule_tail_type _ "match" _ tail_rule_pattern_value {% ([tailType, , matchType, , pattern]) => ({
    tailType,
    matchType,
    pattern,
  }) %}
tail_rule_form ->
  tail_rule_tail_type _ "form" _ tail_rule_form_pattern {% ([tailType, , matchType, , pattern]) => ({
    tailType,
    matchType,
    pattern,
  }) %}
tail_rule_form_pattern -> "{" (_ tail_rule_form_pattern_rule):* _ "}" {% ([, rules]) => rules.map(([comment, rule]) => ({ comment, ...rule })) %}
tail_rule_form_pattern_rule ->
    tail_rule_form_pattern_rule_key {% ([key]) => ({ key, value: null, array: false }) %}
  | tail_rule_form_pattern_rule_key _ "=" _ tail_rule_pattern_value
  {% ([key, /*_*/, /*"="*/, /*_*/, value]) => ({
    key,
    value,
    array: false,
  }) %}
  | tail_rule_form_pattern_rule_key _ "[" _ "]" _ "=" _ tail_rule_pattern_value
  {% ([key, /*_*/, /*[*/, /*_*/, /*]*/, /*_*/, /*"="*/, /*_*/, value]) => ({
    key,
    value,
    array: true,
  }) %}
tail_rule_form_pattern_rule_key ->
    ident {% ([value]) => ({ type: 'id', value }) %}
  | %string {% ([value]) => ({ type: 'string', value }) %}
tail_rule_pattern_value ->
  ("|" _):? (tail_rule_pattern_value_term _ "|" _):* tail_rule_pattern_value_term
  {% ([, rules, rule]) => [...rules.map(([rule]) => rule), rule] %}
tail_rule_pattern_value_term ->
    ident {% ([value]) => ({ type: 'id', value }) %}
  | %string {% ([value]) => ({ type: 'string', value }) %}
  | %regex {% ([value]) => ({ type: 'regex', value }) %}
