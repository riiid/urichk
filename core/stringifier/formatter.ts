import * as ast from "../ast/index.ts";
import {
  Visitor,
  visitor as defaultVisitor,
  visitStatementBase,
} from "../visitor/index.ts";
import { createPrinter } from "./printer.ts";

export default function format(ast: ast.Urichk): string {
  const printer = createPrinter();
  const formatter: Visitor = {
    ...defaultVisitor,
    visitToken(visitor, node) {
      printer.print(node.text);
    },
    visitHead(visitor, node) {
      defaultVisitor.visitHead(visitor, node);
      printer.print(" ");
    },
    visitUserinfo(visitor, node) {
      defaultVisitor.visitUserinfo(visitor, node);
      printer.print("@");
    },
    visitScheme(visitor, node) {
      defaultVisitor.visitScheme(visitor, node);
      printer.print(":");
    },
    visitPort(visitor, node) {
      printer.print(":");
      defaultVisitor.visitPort(visitor, node);
    },
    visitPathFragment(visitor, node) {
      printer.print("/");
      defaultVisitor.visitPathFragment(visitor, node);
    },
    visitParamPathFragment(visitor, node) {
      printer.print("[");
      defaultVisitor.visitParamPathFragment(visitor, node);
      printer.print("]");
    },
    visitTail(visitor, node) {
      visitStatementBase(visitor, node, () => {
        printer.print("{");
        printer.indent();
        for (const rule of node.rules) {
          printer.print("\n");
          visitor.visitTailRule(visitor, rule);
          printer.print("\n");
        }
        printer.dedent();
        printer.println("}\n");
      });
    },
    visitTailRule(visitor, node) {
      defaultVisitor.visitTailRule(visitor, node);
    },
    visitTailRuleMatch(visitor, node) {
      visitStatementBase(visitor, node, () => {
        printer.printIndent();
        const { tailType, matchType, pattern } = node;
        visitor.visitTailType(visitor, tailType);
        visitor.visitMatchType(visitor, matchType);
        if (pattern.length > 1) {
          printer.indent();
          for (const value of pattern) {
            printer.print("\n");
            printer.printIndent();
            printer.print("| ");
            visitor.visitTailRulePatternValue(visitor, value);
          }
          printer.dedent();
        } else {
          for (const value of pattern) {
            visitor.visitTailRulePatternValue(visitor, value);
          }
        }
      });
    },
    visitTailRuleForm(visitor, node) {
      visitStatementBase(visitor, node, () => {
        printer.printIndent();
        const { tailType, matchType, pattern } = node;
        visitor.visitTailType(visitor, tailType);
        visitor.visitMatchType(visitor, matchType);
        printer.print("{\n");
        printer.indent();
        for (const rule of pattern) {
          visitor.visitTailRuleFormPatternRule(visitor, rule);
        }
        printer.dedent();
        printer.printIndent();
        printer.print("}");
      });
    },
    visitTailRuleFormPatternRule(visitor, node) {
      visitStatementBase(visitor, node, () => {
        printer.printIndent();
        const { key, value, array } = node;
        visitor.visitKey(visitor, key);
        if (array) visitor.visitArrayToken(visitor, array);
        if (value) {
          printer.print(" ");
          printer.print("=");
          printer.print(" ");
          value.forEach((patternValue, index) => {
            if (index > 0) printer.print(" | ");
            visitor.visitTailRulePatternValue(visitor, patternValue);
          });
        }
        printer.print("\n");
      });
    },
    visitTailRulePatternValue(visitor, node) {
      defaultVisitor.visitTailRulePatternValue(visitor, node);
    },
    visitArrayToken(visitor, node) {
      printer.print(" ");
      visitor.visitToken(visitor, node);
    },
    visitTailType(visitor, node) {
      defaultVisitor.visitTailType(visitor, node);
      printer.print(" ");
    },
    visitMatchType(visitor, node) {
      defaultVisitor.visitMatchType(visitor, node);
      printer.print(" ");
    },
    visitCommentGroup(visitor, node) {
      printer.printIndent();
      for (const comment of node.comments) {
        visitor.visitComment(visitor, comment);
        printer.print("\n");
      }
    },
    visitComment(visitor, node) {
      if (node.type === "document-comment") {
        node.text.split("\n").map((line, index) => {
          const trimed = line.trimStart();
          if (index) {
            printer.println("");
            printer.printIndent();
            printer.print(" ");
          }
          printer.print(trimed);
        });
      } else {
        visitor.visitToken(visitor, node);
      }
    },
  };
  formatter.visitUrichk(formatter, ast);
  return printer.done();
}
