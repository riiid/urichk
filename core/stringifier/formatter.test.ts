import { parse } from "../parser/urichk.ts";
import format from "./formatter.ts";
import { test } from "https://deno.land/x/snap@0.0.3/mod.ts";

test("#1", ({ assertSnapshot }) => {
  const fixture = `
scheme:example1.com/foo/bar/baz {
  ? match
    /web-path=(?<path>.*)/
  # match id-1
    | id-2
    | /id-3/
}

/**
 * foo
 */
scheme:example2.com/path/[param1]/[param2] {
  /**
   * bar
   */
? form {
    /**
     * baz
     */
    '' = ''
a
    b
    param1 = exact | match | value
  param2 = /^[0-9]+$/
    param3 [] = /^[a-z]+$/i
  }
  # match /id/
}

scheme:username@example3.com:4321 {
  ? form {'a' = /.*/
    'b' = /.*/
    'c-d' = /.*/
    '/' = /.*/
    '?' = /.*/
    'f/' = /.*/
    'g' = /.*/
    'h' = /.*/
  }
}`;

  const parseResult = parse(fixture);
  if (!parseResult) throw new Error("No schema");
  const formatted = format(parseResult.ast);
  assertSnapshot(formatted);
});

test("#2", ({ assertSnapshot }) => {
  const fixture = `
  /**
   * Webview for resetting password of a user.
   */
  /webview/reset-password {
      ? form {
          /**
           * Code sent via email.
           */
          token = /^.+$/
      }
  }
  
  /**
   * Webview for a learning cell.
   */
  /webview/learning-cell {
      ? form {
          /**
           * Learning cell id.
           */
          cellId = /^\d+$/
      }
  }
  
  /**
   * Webview for the coupon box
   */
  /webview/coupon/box {}  
  /*
   * Webview of Virtual Exam Report Page
   */
  /webview/virtual-exam/report {
      ? form {
          /**
           * Session ID (Learning Self Card)
           */
          sessionId = /^\d+$/
          /**
           * It should be 'virtual_exam' if this page loaded after the virtual exam'
           * It should be 'self' if this page loaded by clicking a self card with the ended virtual exam
           */
          referrer = virtual_exam | self
      }
  }
`;

  const parseResult = parse(fixture);
  if (!parseResult) throw new Error("No schema");
  const formatted = format(parseResult.ast);
  assertSnapshot(formatted);
});
