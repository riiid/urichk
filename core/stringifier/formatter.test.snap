{
  '#1': "scheme:example1.com/foo/bar/baz {\n    ? match /web-path=(?<path>.*)/\n\n    # match \n        | id-1\n        | id-2\n        | /id-3/\n}\n\n/**\n * foo\n */\nscheme:example2.com/path/[param1]/[param2] {\n    /**\n     * bar\n     */\n    ? form {\n        /**\n         * baz\n         */\n        '' = ''\n        a\n        b\n        param1 = exact | match | value\n        param2 = /^[0-9]+$/\n        param3 [] = /^[a-z]+$/i\n    }\n\n    # match /id/\n}\n\nscheme:username@example3.com:4321 {\n    ? form {\n        'a' = /.*/\n        'b' = /.*/\n        'c-d' = /.*/\n        '/' = /.*/\n        '?' = /.*/\n        'f/' = /.*/\n        'g' = /.*/\n        'h' = /.*/\n    }\n}\n\n",
  '#2': "/**\n * Webview for resetting password of a user.\n */\n/webview/reset-password {\n    ? form {\n        /**\n         * Code sent via email.\n         */\n        token = /^.+$/\n    }\n}\n\n/**\n * Webview for a learning cell.\n */\n/webview/learning-cell {\n    ? form {\n        /**\n         * Learning cell id.\n         */\n        cellId = /^d+$/\n    }\n}\n\n/**\n * Webview for the coupon box\n */\n/webview/coupon/box {}\n\n/*\n   * Webview of Virtual Exam Report Page\n   */\n/webview/virtual-exam/report {\n    ? form {\n        /**\n         * Session ID (Learning Self Card)\n         */\n        sessionId = /^d+$/\n        /**\n         * It should be 'virtual_exam' if this page loaded after the virtual exam'\n         * It should be 'self' if this page loaded by clicking a self card with the ended virtual exam\n         */\n        referrer = virtual_exam | self\n    }\n}\n\n",
}